python_modules/dagster/dagster/core/definitions/event_metadata/table.py

```diff
@@ -0,0 +1,303 @@
+      `other` is supported. This should contain a list of strings describing
+      arbitrary table-level constraints.
+    - No top-level properties other than `fields` and `constraints` are
+      allowed.
+    - Field descriptors only support `name`, `type`, `description`, and `constraints` properties.
+      `format`, `title`, or arbitrary properties are not allowed.
+    - The `type` of a field descriptor is an arbitrary string (i.e. it is not restricted
+      to Frictionless types).
+    - Field `constraints` descriptor MAY contain a property `other` which
+      contains an array of strings. Each element should describe a constraint
+      that is not expressible with predefined Frictionless constraint types.
+
+    The schema is constructed out of :py:class:`~dagster.TableConstraints` and
+    :py:class:`~dagster.TableField` objects. Example:
+
+    .. code-block:: python
+
+            # example schema
+            TableSchema(
+                constraints = TableConstraints(
+                    "other": [
+                        "foo > bar",
+                    ],
+                ),
+                fields = [
+                    TableField(
+                        name = "foo",
+                        type = "string",
+                        description = "Foo description",
+                        constraints = TableFieldConstraints(
+                            required = True,
+                            other = [
+                                "starts with the letter 'a'",
+                            ],
+                        ),
+                    ),
+                    TableField(
+                        name = "bar",
+                        type = "string",
+                    ),
+                    TableField(
+                        name = "baz",
+                        type = "custom_type",
+                        constraints = TableFieldConstraints(
+                            unique = True,
+                        )
+                    ),
+                ],
+            )
+
+    Args:
+        fields (List[TableField]): The fields of the table.
+        constraints (Optional[TableConstraints]): The constraints of the table.
+    """
+
+    def __new__(
+        cls,
+        fields: List["TableField"],
+        constraints: Optional["TableConstraints"] = None,
+    ):
+        return super(TableSchema, cls).__new__(
+            cls,
+            fields=check.list_param(fields, "fields", of_type=TableField),
+            constraints=check.opt_inst_param(
+                constraints, "constraints", TableConstraints, default=_DefaultTableConstraints
+            ),
+        )
+
+
+# ########################
+# ##### TABLE CONSTRAINTS
+# ########################
+
+
+class _TableConstraintsSerializer(DefaultNamedTupleSerializer):
+    pass
+
+
+@whitelist_for_serdes(serializer=_TableConstraintsSerializer)
+class TableConstraints(
```

Comment:
Are there other properties that we envision supporting? Why use this class instead of just include a list of strings on TableSchema?

Reply:
Because I think we should leave the door open for explicitly formalized table-level constraints (e.g. min/max rows), but wasn't sure what should be included at this time (and Frictionless lacks table-level constraints right now). If we do add these properties later, it is easier to add them as new keys than have to change from a simple list of strings to a dictionary.

Reply:
Ah, I hadn't caught that column/field constraints worked that way. I added a comment on `TableFIeldConstraints`.